view all squids

ohad-porat opened the pull request:
@dpickett PR RFR 

### Accomplished
* Create squids migration, model, and route
* Create SquidList and SquidTile components
* Add `handleErrors` middleware and wrap routes with `nextWrapper.js`
* Sort squids by creation date
* Make the HTTP requests using the `ApiClient` class
* Implement React Query 
* Styling


Loom video: https://www.loom.com/share/b01e495ae6e04593a95fa26d0d34eb11

ohad-porat committed:
WIP view all squids

ohad-porat committed:
es lint error fix

ohad-porat committed:
es lint error fix

ohad-porat committed:
more es lint error fix

ohad-porat committed:
add enum to specialPowers

ohad-porat committed:
WIP seed quids

ohad-porat committed:
WIP add react query

dpickett commented on client/src/components/App.jsx:
```diff
@@ -18,7 +21,9 @@ const App = () => {
-  });
+  })
+
+  // ohad is a total nerd hahahahahaha ULTIMATE POWERRRRRRRRRR!!!!!!!!!
```
lol

dpickett commented on client/src/components/squids/SquidList.jsx:
```diff
@@ -0,0 +1,22 @@
+  const squidListQuery = useSquidList()
+  const squids = squidListQuery.data || []
+  // eslint-disable-next-line no-nested-ternary
+  const squidTiles = squidListQuery.isLoading
```
this ternary is a bit hard to read. I might consider instead breaking out components for each state and conditionally returning them. The variable name also gets a little misleading if it represents text that is not related to squids

dpickett commented on client/src/components/squids/SquidTile.jsx:
```diff
@@ -0,0 +1,15 @@
+
+  return (
+    <div>
+      <p>---</p>
```
you could use a css border to impart this style

dpickett commented on server/src/db/seeders/SquidSeeder.js:
```diff
@@ -0,0 +1,18 @@
+    try {
+      const squidFactory = new Factory(Squid)
+
+      await squidFactory.createMany(15)
```
15 is a lot to create, particularly if. we're seeding in the test suite. It might be a good idea to reduce this count, or make the count conditional on the basis of `NODE_ENV` or another environment variable.

ohad-porat replied:
```javascript
const squidsToCreate = process.env.NODE_ENV === "development" ? 15 : 3
```


```javascript
class Configuration {
  get nodeEnv() {
    return (process.env.NODE_ENV || "") === "" ? "development" : process.env.NODE_ENV
  }
}

const config = new Configuration()
config.nodeEnv
```

dpickett commented on server/src/routes/api/v1/squidsRouter.js:
```diff
@@ -0,0 +1,16 @@
+
+squidsRouter.get(
+  "/",
+  nextWrapper(async (req, res) => {
```
nice! I bet you'll tackle this in the pagination story, but it's a good idea to always constrain our index routes. At a minimium, I'll set a provisional limit to a high number like 1000 or 5000 to eliminate the possibility of the server running out of memory with a large record count.

dpickett reviewed (changes requested):
🚀

ohad-porat committed:
Assimilate react query feedback

WIP fix according to dan's comments

add styling

add link to index page

add fonts, transition on hover

small fix per dan's feedback

ohad-porat commented on server/src/services/Configuration.js:
```diff
@@ -0,0 +1,5 @@
+  get nodeEnv() {
+    return (process.env.NODE_ENV || "") === "" ? "development" : process.env.NODE_ENV
+  }
+}
```
I noticed that we also have a `getNodeEnv.js` service function in the config folder, should I just use that instead?

dpickett commented on client/src/components/App.jsx:
```diff
@@ -7,7 +7,7 @@ import { ReactQueryDevtools } from "react-query/devtools"
 
 import { TopBar } from "./layout/TopBar"
-import { SquidList } from "./squids/SquidList"
+import { SquidList } from "./squidIndex/SquidList"
```
I liked the folder name `squids`

dpickett commented on client/src/components/layout/TopBar.jsx:
```diff
@@ -1,14 +1,22 @@
     <div className="top-bar-left">
       <ul className="menu">
-        <li className="menu-text">App</li>
+        <li className="menu-text">Da Squids App</li>
```
it might be good to have this link to the root

dpickett commented on client/src/components/squidIndex/SquidTile.jsx:
```diff
@@ -0,0 +1,18 @@
+
+export const SquidTile = ({ squid }) => {
+  const specialPowerElement = squid.specialPower ? <p>Special Power: {squid.specialPower}</p> : null
+  const hrLineElement = squid.id === "1" ? "" : <hr />
```
hm - we can't rely on a hard `id` value in production. If we want to separate the tiles I would suggest we do so in the `List` component. 

```javascript
squidListQueryOutput = squids.map((squid) => <div key={squid.id}><SquidTile squid={squid} /><hr /></div>)
```

dpickett commented on client/src/components/squids/SquidList.jsx:
```diff
@@ -3,20 +3,36 @@ import React from "react"
-      <h2>Squid List</h2>
-      {squidTiles}
-    </>
+    <div className="page-body">
```
since this is specific to squids, I would probably call this `squids-list`

dpickett commented on client/src/components/squids/SquidList.jsx:
```diff
@@ -3,20 +3,36 @@ import React from "react"
-      {squidTiles}
-    </>
+    <div className="page-body">
+      <h2 className="index-header__text">Squid List</h2>
```
bem wise, I would call this `squids-list__header`

dpickett commented on client/src/components/squids/SquidTile.jsx:
```diff
@@ -4,9 +4,8 @@ export const SquidTile = ({ squid }) => {
-    <div>
-      <p>---</p>
-      <h3>Name: {squid.name}</h3>
+    <div className="squid-tile">
```
consider: `squids-list__tile`

dpickett commented on client/src/components/squids/SquidTile.jsx:
```diff
@@ -4,9 +4,8 @@ export const SquidTile = ({ squid }) => {
-      <p>---</p>
-      <h3>Name: {squid.name}</h3>
+    <div className="squid-tile">
+      <h3 className="squid-tile__header">Name: {squid.name}</h3>
```
consider `squids-list__tile-header`

dpickett commented on client/src/components/layout/TopBar.jsx:
```diff
@@ -1,14 +1,26 @@
-        <li className="menu-text">App</li>
         <li>
-          <Link to="/">Home</Link>
+          <Link to="/" className="menu-link">
```
BEM wise, this would probably be: `menu__nav-link`

ohad-porat committed:
bem fixes
